### PR TITLE
add PAG support for SD Controlnet Img2Img

### PR DESCRIPTION
**Pull request by Bhavay-2001**
# What does this PR do?

Part of #8710 

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Tagging @yiyixuxu

### CODE REVIEW

**Comment by Bhavay-2001** [comment]
Hi @yiyixuxu, please review this once. I am having some difficulty with the tests so please have a look at that.

File: tests/pipelines/pag/test_pag_controlnet_sd_img2img.py [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+            3,
+        ), f"the shape of the output image should be (1, 64, 64, 3) but got {image.shape}"
+        expected_slice = np.array(
+            [0.66685176, 0.53207266, 0.5541569, 0.5912994, 0.5368312, 0.58433825, 0.42607725, 0.46805605, 0.5098659]
+        )
+
+        max_diff = np.abs(image_slice.flatten() - expected_slice).max()
+        assert max_diff < 1e-3, f"output is different from expected, {image_slice.flatten()}"
+
+    # def test_ip_adapter_single(self):
```
Comment (yiyixuxu) [inline tests/pipelines/pag/test_pag_controlnet_sd_img2img.py:251]: we can remove these tests, no?

Comment (Bhavay-2001) [reply tests/pipelines/pag/test_pag_controlnet_sd_img2img.py:251]: Yes, the commented test will be removed.

Comment (yiyixuxu) [reply tests/pipelines/pag/test_pag_controlnet_sd_img2img.py:251]: can you run `make style` and `make fix-copies` so the quality tests would pass? currently failing

File: src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd_img2img.py [hunk 2]
```diff
@@ -0,0 +1,1327 @@
...
+        return encoder_output.latent_dist.sample(generator)
+    elif hasattr(encoder_output, "latent_dist") and sample_mode == "argmax":
+        return encoder_output.latent_dist.mode()
+    elif hasattr(encoder_output, "latents"):
+        return encoder_output.latents
+    else:
+        raise AttributeError("Could not access latents of provided encoder_output")
+
+
+def prepare_image(image):
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd_img2img.py:111]: Could you add the missing `# Copied from` here?

File: src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd_img2img.py [hunk 3]
```diff
@@ -0,0 +1,1327 @@
...
+        )
+        self.vae_scale_factor = 2 ** (len(self.vae.config.block_out_channels) - 1)
+        self.image_processor = VaeImageProcessor(vae_scale_factor=self.vae_scale_factor, do_convert_rgb=True)
+        self.control_image_processor = VaeImageProcessor(
+            vae_scale_factor=self.vae_scale_factor, do_convert_rgb=True, do_normalize=False
+        )
+        self.register_to_config(requires_safety_checker=requires_safety_checker)
+        self.set_pag_applied_layers(pag_applied_layers)
+
+    # Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion.StableDiffusionPipeline._encode_prompt
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd_img2img.py:240]: I don't think this method is being used anywhere and so can be removed.

File: src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd_img2img.py [hunk 4]
```diff
@@ -0,0 +1,1327 @@
...
+        if accepts_eta:
+            extra_step_kwargs["eta"] = eta
+
+        # check if the scheduler accepts generator
+        accepts_generator = "generator" in set(inspect.signature(self.scheduler.step).parameters.keys())
+        if accepts_generator:
+            extra_step_kwargs["generator"] = generator
+        return extra_step_kwargs
+
+    def check_inputs(
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd_img2img.py:572]: Please add the missing `# Copied from` here as well

File: src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd_img2img.py [hunk 5]
```diff
@@ -0,0 +1,1327 @@
...
+
+        Returns:
+            [`~pipelines.stable_diffusion.StableDiffusionPipelineOutput`] or `tuple`:
+                If `return_dict` is `True`, [`~pipelines.stable_diffusion.StableDiffusionPipelineOutput`] is returned,
+                otherwise a `tuple` is returned where the first element is a list with the generated images and the
+                second element is a list of `bool`s indicating whether the corresponding generated image contains
+                "not-safe-for-work" (nsfw) content.
+        """
+
+        callback = kwargs.pop("callback", None)
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd_img2img.py:1038]: These callbacks have been deprecated. You can remove them

File: src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd_img2img.py [hunk 6]
```diff
@@ -0,0 +1,1327 @@
...
+        control_guidance_start: Union[float, List[float]] = 0.0,
+        control_guidance_end: Union[float, List[float]] = 1.0,
+        clip_skip: Optional[int] = None,
+        callback_on_step_end: Optional[
+            Union[Callable[[int, int, Dict], None], PipelineCallback, MultiPipelineCallbacks]
+        ] = None,
+        callback_on_step_end_tensor_inputs: List[str] = ["latents"],
+        pag_scale: float = 3.0,
+        pag_adaptive_scale: float = 0.0,
+        **kwargs,
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd_img2img.py:927]:
```suggestion
```

File: src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd_img2img.py [hunk 7]
```diff
@@ -0,0 +1,1327 @@
...
+                "callback",
+                "1.0.0",
+                "Passing `callback` as an input argument to `__call__` is deprecated, consider using `callback_on_step_end`",
+            )
+        if callback_steps is not None:
+            deprecate(
+                "callback_steps",
+                "1.0.0",
+                "Passing `callback_steps` as an input argument to `__call__` is deprecated, consider using `callback_on_step_end`",
+            )
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd_img2img.py:1052]:
```suggestion
```

File: src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd_img2img.py [hunk 8]
```diff
@@ -0,0 +1,1327 @@
...
+                    guess_mode=guess_mode,
+                )
+
+                control_images.append(control_image_)
+
+            control_image = control_images
+        else:
+            assert False
+
+        # 5. Prepare timesteps
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd_img2img.py:1181]: You might have to fix the step numbering here

File: src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd_img2img.py [hunk 9]
```diff
@@ -0,0 +1,1327 @@
...
+                for s, e in zip(control_guidance_start, control_guidance_end)
+            ]
+            controlnet_keep.append(keeps[0] if isinstance(controlnet, ControlNetModel) else keeps)
+
+        # 8. Denoising loop
+        num_warmup_steps = len(timesteps) - num_inference_steps * self.scheduler.order
+        with self.progress_bar(total=num_inference_steps) as progress_bar:
+            for i, t in enumerate(timesteps):
+                # expand the latents if we are doing classifier free guidance
+                latent_model_input = torch.cat([latents] * 2) if self.do_classifier_free_guidance else latents
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd_img2img.py:1223]: Are you sure the forward pass is working? Shouldn't this be something like https://github.com/huggingface/diffusers/blob/98388670d293a590e69d9ff6b442169829f47cf3/src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd_xl.py#L1492

File: src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd_img2img.py [hunk 10]
```diff
@@ -0,0 +1,1327 @@
...
+                    latents = callback_outputs.pop("latents", latents)
+                    prompt_embeds = callback_outputs.pop("prompt_embeds", prompt_embeds)
+                    negative_prompt_embeds = callback_outputs.pop("negative_prompt_embeds", negative_prompt_embeds)
+
+                # call the callback, if provided
+                if i == len(timesteps) - 1 or ((i + 1) > num_warmup_steps and (i + 1) % self.scheduler.order == 0):
+                    progress_bar.update()
+                    if callback is not None and i % callback_steps == 0:
+                        step_idx = i // getattr(self.scheduler, "order", 1)
+                        callback(step_idx, t, latents)
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd_img2img.py:1296]:
```suggestion
```

File: src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd_img2img.py [hunk 11]
```diff
@@ -0,0 +1,1327 @@
...
+                    encoder_hidden_states=prompt_embeds,
+                    cross_attention_kwargs=self.cross_attention_kwargs,
+                    down_block_additional_residuals=down_block_res_samples,
+                    mid_block_additional_residual=mid_block_res_sample,
+                    added_cond_kwargs=added_cond_kwargs,
+                    return_dict=False,
+                )[0]
+
+                # perform guidance
+                if self.do_classifier_free_guidance:
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd_img2img.py:1274]: Perturbed guidance part does not seem to have been implemented at the different places where it is supposed to be added.

File: src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd_img2img.py [hunk 12]
```diff
@@ -0,0 +1,1327 @@
...
+            clip_skip=self.clip_skip,
+        )
+        # For classifier free guidance, we need to do two forward passes.
+        # Here we concatenate the unconditional and text embeddings into a single batch
+        # to avoid doing two forward passes
+        if self.do_classifier_free_guidance:
+            prompt_embeds = torch.cat([negative_prompt_embeds, prompt_embeds])
+
+        if ip_adapter_image is not None or ip_adapter_image_embeds is not None:
+            image_embeds = self.prepare_ip_adapter_image_embeds(
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd_img2img.py:1135]: IP Adapter perturbed embeddings need to be generated differently. Please refer to one of the linked PRs.

**Review by a-r-r-o-w — CHANGES_REQUESTED** [review]
Thank you for adding support for this! It seems like a lot of the perturbed attention guidance part is not handled yet and this is a raw copy ControlNetSDImg2Img, no? Let me try and help you with the required changes that need to be made:
- Refer to #7944 and #8725
- Try and understand what happens in [pag_utils.py](https://github.com/huggingface/diffusers/blob/98388670d293a590e69d9ff6b442169829f47cf3/src/diffusers/pipelines/pag/pag_utils.py) file.
- Take a look at one of the existing PAG pipeline implementations. Notice that at various locations in the code, we check the `self.do_perturbed_attention_guidance` flag and handle things differently from normal CFG. You will have to apply these changes as well. It might be a little tricky to do with controlnet since you also need to take care of `control_model_input`. The easiest way to see all the differences would be to view the diff of non-PAG and PAG variants (for example, StableDiffusionXLPipeline and StableDiffusionXLPAGPipeline) side-by-side
- Once you're comfortable and have made all the required changes, try and run through all the different scenarios such as `guess_mode` true and false, with guidance_scale == 1 and guidance_scale > 1, pag_scale == 0 and pag_scale > 0, etc.

If you need any additional help, feel free to ping me any time.

**Comment by a-r-r-o-w** [comment]
Also, it might help to change the title of this PR to something like "add PAG support for SD Controlnet Img2Img" to reflect the intent correctly when merged

File: src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd_img2img.py (hunk 12 above)
Comment (Bhavay-2001) [reply src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd_img2img.py:1135]: Hi @a-r-r-o-w, are you referring to this [part](https://github.com/huggingface/diffusers/blob/673eb60f1c4d971e1a577bed767053e50578b461/src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd_xl.py#L1370) here?

Comment (a-r-r-o-w) [reply src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd_img2img.py:1135]: No, I'm referring to this: https://github.com/huggingface/diffusers/blob/673eb60f1c4d971e1a577bed767053e50578b461/src/diffusers/pipelines/pag/pipeline_pag_sd_xl.py#L1156

**Comment by Bhavay-2001** [comment]
Hi @a-r-r-o-w, I am trying to fix the coding mistakes and coding style by calling the `make style` and other similar commands but it says make command not found. I know there is something wrong but cannot figure out. Could you pls explain that once how its done?

**Comment by a-r-r-o-w** [comment]
You will need to install make. I'm assuming you are on Windows since make should be available by default on linux or mac. If you install Git for Windows, you will easily be able to use it. Otherwise, try https://stackoverflow.com/questions/32127524/how-to-install-and-use-make-in-windows

**Comment by Bhavay-2001** [comment]
Hi @a-r-r-o-w, I am first completing the pag pipeline and not the associated tests. Please check this whenever you have time. Thanks

**Comment by a-r-r-o-w** [comment]
Can you show us some examples with and without PAG enabled? Also, please post the minimal reproducible example. And yes, we can work on tests later in a similar fashion to how others have done in their PRs.

**Comment by Bhavay-2001** [comment]
Hi, I am shifting this PR [here](https://github.com/huggingface/diffusers/pull/8864). Further communications will be done there. I am closing this one. 
Thanks

**Pull request closed without merging**